sqldelight-idea-plugin/src/main/kotlin/com/squareup/sqldelight/intellij/actions/file/SqlDelightCreateFileAction.kt

```diff
@@ -0,0 +1,186 @@
+import com.intellij.openapi.application.runWriteAction
+import com.intellij.openapi.editor.LogicalPosition
+import com.intellij.openapi.fileEditor.FileEditorManager
+import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.project.DumbService
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.InputValidatorEx
+import com.intellij.openapi.vfs.VfsUtil
+import com.intellij.psi.PsiDirectory
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.testFramework.LightVirtualFile
+import com.intellij.util.IncorrectOperationException
+import com.squareup.sqldelight.core.SqlDelightPropertiesFile
+import com.squareup.sqldelight.core.lang.SqlDelightFile
+import com.squareup.sqldelight.core.lang.SqlDelightFileType
+import com.squareup.sqldelight.core.lang.SqlDelightLanguage
+import com.squareup.sqldelight.intellij.SqlDelightFileIconProvider
+import com.squareup.sqldelight.intellij.propertiesFolder
+import java.util.Properties
+
+/**
+ * Creates a new SqlDelight file/table/migration from a template (see [fileTemplates.internal])
+ *
+ * Adopted from [NewKotlinFileAction](https://github.com/JetBrains/kotlin/blob/f6a739bbc5b66a8f2cb5cb61acb1e87c899fa11e/idea/src/org/jetbrains/kotlin/idea/actions/NewKotlinFileAction.kt#L56)
+ */
+class SqlDelightCreateFileAction : CreateFileFromTemplateAction(
+    CAPTION,
+    "Creates new SqlDelight file or table",
+    SqlDelightFileType.icon
+), DumbAware {
+
+  override fun postProcess(
+    createdElement: PsiFile?,
+    templateName: String?,
+    customProperties: MutableMap<String, String>?
+  ) {
+    super.postProcess(createdElement, templateName, customProperties)
+
+    if (createdElement is SqlDelightFile) {
+      val isTableTemplate = PsiTreeUtil.collectElements(createdElement) { it is SqlTableName }.isNotEmpty()
+      val editor = FileEditorManager.getInstance(createdElement.project).selectedTextEditor ?: return
+      if (editor.document == createdElement.viewProvider.document) {
+        if (isTableTemplate) {
+          // move caret inside "create table" sql statement
+          val lineCount = editor.document.lineCount
+          if (lineCount > 0) {
+            editor.caretModel.moveToLogicalPosition(LogicalPosition(lineCount - 2, 0))
+          }
+        } else {
+          // move caret after "create table", so the user can enter the table name conveniently
+          val tableKeyword = PsiTreeUtil.collectElements(createdElement) { it.node.elementType == TABLE }.first()
+          editor.caretModel.moveToOffset(tableKeyword.textRange.endOffset + 1)
+        }
+      }
+    }
+  }
+
+  override fun getActionName(directory: PsiDirectory?, newName: String, templateName: String?): String = CAPTION
+
+  override fun buildDialog(project: Project?, directory: PsiDirectory?, builder: CreateFileFromTemplateDialog.Builder) {
+    val icon = iconProvider.getIcon(dummyFile, 0, project)
```

Comment:
i like using the iconProvider here, it's more accurate to the file that's about to be created and the icon it will have

Reply:
i think we should add an api to `SqlDelightFileIconProvider` that returns the icon for a language/project, and then that way you don't need the dummyFile type

Reply:
Good point, will do 👍